File: Source/SafetySharp/CSharp/Normalization/CSharpNormalizer.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.CSharp.Normalization
{
	using System;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp;
	using Utilities;

	/// <summary>
	///     A base class for C# normalizers that normalize certain C# features to equivalent ones that are easier to transform to
	///     the metamodel.
	/// </summary>
	public abstract class CSharpNormalizer : CSharpSyntaxRewriter
	{
		/// <summary>
		///     The semantic model that can be used for semantic analysis during normalization.
		/// </summary>
		protected SemanticModel SemanticModel { get; private set; }

		/// <summary>
		///     Normalizes the C# code contained in <paramref name="compilation." />
		/// </summary>
		/// <param name="compilation">The C# compilation that should be normalized.</param>
		public Compilation Normalize(Compilation compilation)
		{
			foreach (var syntaxTree in compilation.SyntaxTrees)
			{
				SemanticModel = compilation.GetSemanticModel(syntaxTree);

				var root = syntaxTree.GetRoot();
				var normalizedRoot = Visit(root);

				compilation = compilation.ReplaceSyntaxTree(syntaxTree, SyntaxFactory.SyntaxTree(normalizedRoot));
			}

			return compilation;
		}

		/// <summary>
		///     Applies all C# code normalizers to the C# <paramref name="compilation" />
		/// </summary>
		/// <param name="compilation">The C# compilation that should be normalized.</param>
		public static Compilation ApplyNormalizers(Compilation compilation)
		{
			Argument.NotNull(compilation, () => compilation);

			compilation = ApplyNormalizer<SafetySharpTypesNormalizer>(compilation);
			compilation = ApplyNormalizer<TriviaNormalizer>(compilation);
			compilation = ApplyNormalizer<ChooseNormalizer>(compilation);

			return compilation;
		}

		/// <summary>
		///     Applies the normalizer of type <typeparamref name="TNormalizer" /> to the C# <paramref name="compilation" />.
		/// </summary>
		/// <typeparam name="TNormalizer">The type of the normalizer that should be applied.</typeparam>
		/// <param name="compilation">The C# compilation that should be normalized.</param>
		private static Compilation ApplyNormalizer<TNormalizer>(Compilation compilation)
			where TNormalizer : CSharpNormalizer, new()
		{
			var normalizer = new TNormalizer();
			return normalizer.Normalize(compilation);
		}
	}
}

File: Source/SafetySharp/CSharp/Normalization/CSharpNormalizer.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.Internal.CSharp.Normalization

open System
open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp
open SafetySharp.Internal.CSharp.Roslyn

/// Indicates which parts of the code are affected by a normalizer.
[<RequireQualifiedAccess>]
type internal NormalizationScope =
    /// Limits the scope of the normalizer to all members of a component class.
    | Components
    /// Limits the scope of the normalizer to all statements (excluding those of the constructors) of a component class.
    | ComponentStatements
    /// Does not limit the scope of the normalizer.
    | Global

/// A base class for C# normalizers that normalize certain C# features to equivalent ones that are easier to transform to
/// the metamodel.
[<AbstractClass>]
type internal CSharpNormalizer (scope) =
    inherit CSharpSyntaxRewriter ()

    /// The semantic model that should be used for semantic analysis during normalization.
    [<DefaultValue>] val mutable semanticModel : SemanticModel

    /// Normalizes the given syntax tree of the compilation.
    member private this.NormalizeSyntaxTree (compilation : Compilation) (syntaxTree : SyntaxTree) =
        this.semanticModel <- compilation.GetSemanticModel syntaxTree

        let root = syntaxTree.GetRoot ()
        let normalizedRoot = this.Visit root

        if root = normalizedRoot then
            syntaxTree
        else
            syntaxTree.WithChangedText (normalizedRoot.GetText ())

    /// Normalizes the C# code contained in <paramref name="compilation." />
    member this.Normalize (compilation : Compilation) =
        compilation.SyntaxTrees 
        |> Seq.fold (fun (compilation : Compilation) syntaxTree -> 
            compilation.ReplaceSyntaxTree (syntaxTree, this.NormalizeSyntaxTree compilation syntaxTree)
        ) compilation

    /// Ensures that non-component classes are only visited when the normalizer as global scope.
    override this.VisitClassDeclaration node =
        if scope = NormalizationScope.Global || node.IsComponentDeclaration this.semanticModel then
            base.VisitClassDeclaration node
        else
            upcast node

    /// Ensures that a constructor is only visited when the normalizer as global scope.
    override this.VisitConstructorDeclaration node =
        if scope <> NormalizationScope.ComponentStatements then
            base.VisitConstructorDeclaration node
        else
            upcast node